rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1175 @@
+        &mut self, min_tx_id: u64, state: &mut u64,
+    ) -> Option<(&HTTP2Transaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+// C exports.
+
+export_tx_get_detect_state!(rs_http2_tx_get_detect_state, HTTP2Transaction);
+export_tx_set_detect_state!(rs_http2_tx_set_detect_state, HTTP2Transaction);
+
+export_tx_detect_flags_set!(rs_http2_set_tx_detect_flags, HTTP2Transaction);
+export_tx_detect_flags_get!(rs_http2_get_tx_detect_flags, HTTP2Transaction);
+
+//TODOnext connection upgrade from HTTP1 cf SMTP STARTTLS
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_http2_probing_parser_tc(
+    _flow: *const Flow, _direction: u8, input: *const u8, input_len: u32, _rdir: *mut u8,
+) -> AppProto {
+    if input != std::ptr::null_mut() {
+        let slice = build_slice!(input, input_len as usize);
+        match parser::http2_parse_frame_header(slice) {
+            Ok((_, header)) => {
+                if header.reserved != 0
+                    || header.length > HTTP2_DEFAULT_MAX_FRAME_SIZE
+                    || header.flags & 0xFE != 0
+                    || header.ftype != parser::HTTP2FrameType::SETTINGS as u8
+                {
+                    return unsafe { ALPROTO_FAILED };
+                }
+                return unsafe { ALPROTO_HTTP2 };
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                return ALPROTO_UNKNOWN;
+            }
+            Err(_) => {
+                return unsafe { ALPROTO_FAILED };
+            }
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_new() -> *mut std::os::raw::c_void {
+    let state = HTTP2State::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_free(state: *mut std::os::raw::c_void) {
+    // Just unbox...
+    let mut state: Box<HTTP2State> = unsafe { transmute(state) };
+    state.free();
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_tx_free(state: *mut std::os::raw::c_void, tx_id: u64) {
+    let state = cast_pointer!(state, HTTP2State);
+    state.free_tx(tx_id);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_parse_ts(
+    flow: *const Flow, state: *mut std::os::raw::c_void, _pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    let state = cast_pointer!(state, HTTP2State);
+    let mut buf = build_slice!(input, input_len as usize);
+
+    if state.progress < HTTP2ConnectionState::Http2StateMagicDone {
```

Comment:
think this should be moved to a non-unsafe function. The FFI funcs should be a simple as possible: convert unsafe input into rust types, call safe rust funcs